JoinColony/colonyCDapp
# Port Metatransactions

This PR adjusts and fixes _(more than actually porting)_ metatransactions to work within the CDapp environment.

Overview of what this PR does:
- update the transaction encoding method _(due to the Ethers upgrade from the Dapp)_
- fixes ganache web3-onboard module message signing _(so that metatransactions work in dev mode, Metamask had not issue)_
- refactor sagas that used "withProofs" contract call helpers, to use the actual methods and provide own proofs _(the "withProofs" methods are not "real" contract methods, but helpers introduced by `colonyJS`)_
- minor fixes

### Issues found
- Setting permissions is flaky and sometimes the action doesn't show up in the list #1008 
- #1009

### Changes:
- Fixed `ganacheWalletModule` personal signature method
- Fixed `getMetatransactionPromise` to update the function encode method
- Created proofs saga util (for `getMoveFundsPermissionProofs` and `getMultiPermissionProofs`)
- Updated the add domain saga to not use the "withProofs" method
- Updated the edit domain saga to not use the "withProofs" method
- Updated the create payment saga to not use the "withProofs" method
- Updated the move funds saga to not use the "withProofs" method
- Updated the manage permissions saga to not use the "withProofs" method
- Updated the smite saga to not use the "withProofs" method
- Fixed the create colony saga to actually work for metatransactions
- Refactored manage permissions saga to use `intArrayToBytes32` for generating hex permissions string
- Updated extension enable saga to not use "withProofs" method
- Fixed `getMetatransactionPromise` to properly set the token client's `metatransactionVariation` _(this was done due to a bug being identified and fixed on the Dapp side https://github.com/JoinColony/colonyDapp/commit/56e605d7171d6b6e33ab2430a1a8ef140ecb7362)_
- Refactor the stake motion saga to not use "withProofs" method
- Refactor the claim motion saga to not use "withProofs" method
- Refactor the claim all motion saga to not use "withProofs" method
- Refactor the escalate motion saga to not use "withProofs" method _(even though this is not really wired in properly, because a lack of UI/feature specs)_

### Docs

**[Add Metatransactions Documentation](https://github.com/JoinColony/ColonyFrontEndLivingStandard/pull/20)**

### Testing

Execute one _(or preferably all)_ the following transactions, with _(and without)_ metatransactions enabled:
- Add Domain Action
- Edit Domain Action
- Create Payment Action
- Move Funds Between Domains Action
- Manage Permissions Action _(Note the bug I've mentioned above, which will make this harder to test, see #1008 )_
- Smite Action
- Create a colony
- Extension Enable
- Once you have a Motion created, test Staking said motion
- Once you have a Motion finalized, test Claiming motion winnings/losses
- Once you have a Motion finalized, test Claiming All motions winnings/losses _(from the user token activation menu)_

_Note: that while I refactored the motion escalate saga as well, that was never wired properly in the Dapp and since it was ported that way, it cannot be tested._

### Demo

_The call will be scheduled once I'm back from my time off_
